feature/experimental/p4rt/ate_tests/lldp_packetin_test/lldp_packetin_test.go

```diff
@@ -0,0 +1,502 @@
+	metadataEgressPort  = *ygot.Uint32(2)
+)
+
+var (
+	dutPort1 = attrs.Attributes{
+		Desc:    "dutPort1",
+		IPv4:    "192.0.2.1",
+		IPv4Len: ipv4PrefixLen,
+	}
+
+	atePort1 = attrs.Attributes{
+		Name:    "atePort1",
+		IPv4:    "192.0.2.2",
+		IPv4Len: ipv4PrefixLen,
+	}
+
+	dutPort2 = attrs.Attributes{
+		Desc:    "dutPort2",
+		IPv4:    "192.0.2.5",
+		IPv4Len: ipv4PrefixLen,
+	}
+
+	atePort2 = attrs.Attributes{
+		Name:    "atePort2",
+		IPv4:    "192.0.2.6",
+		IPv4Len: ipv4PrefixLen,
+	}
+)
+
+type PacketIO interface {
+	GetTableEntry(delete bool) []*p4rtutils.ACLWbbIngressTableEntryInfo
+	GetPacketTemplate() *PacketIOPacket
+	GetTrafficFlow(ate *ondatra.ATEDevice, frameSize uint32, frameRate uint64) []*ondatra.Flow
+	GetEgressPort() []string
+	GetIngressPort() string
+}
+
+type PacketIOPacket struct {
+	SrcMAC, DstMAC *string
+	EthernetType   *uint32
+}
+
+type testArgs struct {
+	ctx      context.Context
+	leader   *p4rt_client.P4RTClient
+	follower *p4rt_client.P4RTClient
+	dut      *ondatra.DUTDevice
+	ate      *ondatra.ATEDevice
+	top      *ondatra.ATETopology
+	packetIO PacketIO
+}
+
+// programmTableEntry programs or deletes p4rt table entry based on delete flag.
+func programmTableEntry(ctx context.Context, t *testing.T, client *p4rt_client.P4RTClient, packetIO PacketIO, delete bool) error {
+	t.Helper()
+	err := client.Write(&p4v1.WriteRequest{
+		DeviceId:   deviceID,
+		ElectionId: &p4v1.Uint128{High: uint64(0), Low: electionID},
+		Updates: p4rtutils.ACLWbbIngressTableEntryGet(
+			packetIO.GetTableEntry(delete),
+		),
+		Atomicity: p4v1.WriteRequest_CONTINUE_ON_ERROR,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// decodePacket decodes L2 header in the packet and returns destination MAC and ethernet type.
+func decodePacket(t *testing.T, packetData []byte) (string, layers.EthernetType) {
+	t.Helper()
+	packet := gopacket.NewPacket(packetData, layers.LayerTypeEthernet, gopacket.Default)
+	etherHeader := packet.Layer(layers.LayerTypeEthernet)
+	if etherHeader != nil {
+		header, decoded := etherHeader.(*layers.Ethernet)
+		if decoded {
+			return header.DstMAC.String(), header.EthernetType
+		}
+	}
+	return "", layers.EthernetType(0)
+}
+
+// testTraffic sends traffic flow for duration seconds.
+func testTraffic(t *testing.T, ate *ondatra.ATEDevice, flows []*ondatra.Flow, srcEndPoint *ondatra.Interface, duration int) {
+	t.Helper()
+	for _, flow := range flows {
+		flow.WithSrcEndpoints(srcEndPoint).WithDstEndpoints(srcEndPoint)
+	}
+	ate.Traffic().Start(t, flows...)
+	time.Sleep(time.Duration(duration) * time.Second)
+
+	ate.Traffic().Stop(t)
+}
+
+// fetchPackets reads p4rt packets sent to p4rt client.
+func fetchPackets(ctx context.Context, t *testing.T, client *p4rt_client.P4RTClient, expectNumber int) []*p4rt_client.P4RTPacketInfo {
+	t.Helper()
+	packets := []*p4rt_client.P4RTPacketInfo{}
+	for i := 0; i < expectNumber; i++ {
+		_, packet, err := client.StreamChannelGetPacket(&streamName, 0)
```

Comment:
Is `minSeqNum==0` a correct value here?

Do we not need to wait for atleast 1 packet or more?

Also can you add `StreamTermErr` here as well after `err` in `StreamChannelGetPacket`?

Reply:
any non-zero value for `minSeqNum`  will make `StreamChannelGetPacket` wait for that many packets. 0 values makes it asynchronous.

Reply:
Based on the `GetPacket()` code from the client, we would not wait for packets if `minSeqNum` is 0. But, by the time this loop completes: `for i := 0; i < expectNumber; i++ {}` we may not have all the packets in the Queue and the test may fail. So every time we need to wait for 1 packet in the queue right?

```
func (p *P4RTClientStream) GetPacket(minSeqNum uint64) (uint64, *P4RTPacketInfo, error) {
	p.pkt_mu.Lock()
	defer p.pkt_mu.Unlock()

	for p.pktCounters.RxPktCntr < minSeqNum {
		if glog.V(2) {
			glog.Infof("'%s' Waiting on Packet (%d/%d)\n",
				p, p.pktCounters.RxPktCntr, minSeqNum)
		}
		if p.ShouldStop() {
			return 0, nil, io.EOF
		}
		p.pktCond.Wait()
	}

	if len(p.pktQ) == 0 {
		return p.pktCounters.RxPktCntr, nil, nil
	}

	pktInfo := p.pktQ[0]
	p.pktQ = p.pktQ[1:]

	return p.pktCounters.RxPktCntr, pktInfo, nil
}
```

Reply:
You're right in that we might not have all packets received by the time the loop completes (although unlikely, since traffic is stopped and we wait some time before calling fetchPackets). The problem, however, is that in the event of a packet loss, the test would just hangs if we use any value other than 0. So instead, I updated fetchPacket to iterate for some duration of time rather than a fixed count.

On a related note, the test was checking if at least one packet is received and considering that a pass. This is because 'expectNumber' was estimated from the rate and duration. I've updated that to querry tgen for the tx packet count and the test will now only pass if all transmitted packets are received.

Reply:
Checked offline with Ankur, reoslving this comment